src/services/ServiceTypes.ts

```diff
@@ -0,0 +1,211 @@
+import type { default as prismaGlobal } from '@/prisma'
+import type { z } from 'zod'
+import type { PrismaClient } from '@prisma/client'
+import type { Auther } from '@/auth/auther/Auther'
+import type { SessionMaybeUser } from '@/auth/Session'
+
+export type PrismaTransaction = Parameters<Parameters<typeof prismaGlobal.$transaction>[0]>[0]
+
+export type TypeValidateReturn<
+    DetailedType
+> = {
+    success: true,
+    data: DetailedType,
+} | {
+    success: false,
+    error: z.ZodError,
+}
+
+export type PrismaPossibleTransaction<
+    WantsToOpenTransaction extends boolean
+> = WantsToOpenTransaction extends true ? PrismaClient : PrismaTransaction
+
+export type PrismaTransactionOrNew<
+    WantsToOpenTransaction extends boolean
+> = PrismaPossibleTransaction<WantsToOpenTransaction> | 'NEW'
+
+
+export type AuthRunConfig = {
+    withAuth: boolean,
+}
+
+type ConfigType<Params> = {
+    params: Params,
+    session: SessionMaybeUser | null
+}
+
+type ConfigTypeData<DetailedType, Params> = ConfigType<Params> & {
+    data: DetailedType,
+}
+
+export type Execute<
+    WithValidation extends boolean,
+    DetailedType,
+    Params,
+    Return,
+    WithAuthParam extends boolean
+> = WithAuthParam extends true ? {
+    execute: (config: WithValidation extends true
+        ? ConfigTypeData<DetailedType, Params>
+        : ConfigType<Params>,
+        authRunConfig: AuthRunConfig) => Promise<Return>
+} : {
+    execute: (config: WithValidation extends true
+        ? ConfigTypeData<DetailedType, Params>
+        : ConfigType<Params>) => Promise<Return>
+}
+
+export type Client<
+    WithValidation extends boolean,
+    DetailedType,
+    Params,
+    Return,
+    WithAuthParam extends boolean,
+    WantsToOpenTransaction extends boolean
+> = {
+    client: (
+        prisma: PrismaTransactionOrNew<WantsToOpenTransaction>
```

Comment:
The `'NEW'` keyword works, but is it easier to just accept zero arguments, and then use a new prisma client?

Reply:
I belive NEW is more clear. It's semantics, I think passing null or nothing is a little implicit.